crates/accelerate/src/twirling.rs

```diff
@@ -0,0 +1,384 @@
+                            let new_block = generate_twirled_circuit(
+                                py,
+                                block,
+                                rng,
+                                twirling_mask,
+                                run_pass,
+                                optimizer_target,
+                                optimizer_global_decomposer,
+                                optimizer_basis_gates,
+                            )?;
+                            Ok(new_block.into_py(py))
+                        })
+                        .collect();
+                    let new_blocks = new_blocks?;
+                    let blocks_list = PyList::new_bound(
+                        py,
+                        new_blocks.iter().map(|block| {
+                            QUANTUM_CIRCUIT
+                                .get_bound(py)
+                                .call_method1(intern!(py, "_from_circuit_data"), (block,))
+                                .unwrap()
+                        }),
+                    );
+
+                    let new_inst_obj = py_inst
+                        .instruction
+                        .bind(py)
+                        .call_method1(intern!(py, "replace_blocks"), (blocks_list,))?
+                        .unbind();
+                    let new_inst = PyInstruction {
+                        qubits: py_inst.qubits,
+                        clbits: py_inst.clbits,
+                        params: py_inst.params,
+                        op_name: py_inst.op_name.clone(),
+                        control_flow: true,
+                        instruction: new_inst_obj.clone_ref(py),
+                    };
+                    let new_inst = PackedInstruction {
+                        op: PackedOperation::from_instruction(Box::new(new_inst)),
+                        qubits: inst.qubits,
+                        clbits: inst.clbits,
+                        params: Some(Box::new(
+                            new_blocks
+                                .iter()
+                                .map(|x| Param::Obj(x.into_py(py)))
+                                .collect::<SmallVec<[Param; 3]>>(),
+                        )),
+                        extra_attrs: inst.extra_attrs.clone(),
+                        #[cfg(feature = "cache_pygates")]
+                        py_op: std::cell::OnceCell::new(),
+                    };
+                    #[cfg(feature = "cache_pygates")]
+                    new_inst.py_op.set(new_inst_obj).unwrap();
+                    out_circ.push(py, new_inst)?;
+                } else {
+                    out_circ.push(py, inst.clone())?;
+                }
+            }
+            _ => {
+                out_circ.push(py, inst.clone())?;
+            }
+        }
+    }
+    if run_pass {
+        let mut dag = DAGCircuit::from_circuit_data(py, out_circ, false)?;
+        optimize_1q_gates_decomposition(
```

Comment:
I opted to just call `Optimize1qGatesDecomposition` from rust here. You could potentially avoid scanning the entire dag to search for runs, because you can search from where we inject the pauli gates to find new runs created by the insertions, but the do that efficiently we'd have to convert to a dag anyway and the point we're doing that it doesn't make a huge difference and it's a lot simpler to just call the existing function.